include/wil/com.h

```diff
@@ -3325,6 +3325,79 @@ WI_NODISCARD auto make_range(IEnumXxx* enumPtr)
 #endif // WIL_HAS_CXX_17
 #endif // WIL_ENABLE_EXCEPTIONS
 
+#if WINAPI_FAMILY_PARTITION(WINAPI_PARTITION_DESKTOP | WINAPI_PARTITION_SYSTEM)
+#ifdef __WIL_WIN32_HELPERS_INCLUDED
+
+/** RAII support for making cross-apartment (or cross process) COM calls with a timeout applied to them.
+ * When this is active any timed out calls will fail with an RPC error code such as RPC_E_CALL_CANCELED.
+ * This is a shared timeout that applies to all calls made on the current thread for the lifetime of
+ * the wil::rpc_timeout object.
+~~~
+{
+  auto timeout = wil::rpc_timeout(5000);
+  remote_object->BlockingCOMCall();
+  remote_object->AnotherBlockingCOMCall();

```

Comment:
This accidentally illustrates one of the subtle gotchas of this sort of timeout: it causes the cancelation of a single topmost call, and if you don't bail out after that 1 failure and make a different hanging call (or if cancelation happens to land while you're between calls, e.g. in a loop), then this will not save you.

Presumably in this comment, remote_object is a C++/WinRT-style wrapper that will throw if BlockingCOMCall returns RPC_E_CALL_CANCELED. But if somebody naively does this same thing without the throwing wrapper, then if both calls hang, the timeout will help with the first and won't apply to the second.

Reply:
I realized this after sending the PR and was not able to finish fixing it up before I left for the day.  The previous internal wrapper used a periodic timer to catch additional calls if the first call did not early-return.  I am bringing that back to ensure that it covers any number of calls.  However, when I added the additional test case for that I encountered a problem that I haven't been able to address just yet.

Reply:
> or if cancelation happens to land while you're between calls

The single failing call seems easy enough for consumers to handle themselves (and is arguably good COM hygiene), but this part is a little scary. I second the persistence of the periodic timer